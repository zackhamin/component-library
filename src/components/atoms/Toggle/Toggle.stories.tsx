import { Story } from '@storybook/react';
import { Toggle, ToggleProps } from './Toggle';

export default {
	title: 'atoms/Toggle',
	component: Toggle,
};

const Template: Story<ToggleProps> = (args) => <Toggle {...args} />;

const defaultProps = {
	children: 'Toggle',
	id: 'toggle',
};

export const Default = Template.bind({});
Default.args = {
	...defaultProps,
};

export const Disabled = Template.bind({});
Disabled.args = {
	...defaultProps,
	isDisabled: true,
};

export const ToggleChecked = Template.bind({});
ToggleChecked.args = {
	...defaultProps,
	isChecked: true,
};

export const ToggleRequired = Template.bind({});
ToggleRequired.args = {
	...defaultProps,
	isRequired: true,
};
