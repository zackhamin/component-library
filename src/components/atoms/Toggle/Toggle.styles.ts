import styled from 'styled-components';
import { colors, typography, visuallyHiddenStyles } from 'Styles';

const StyledToggleLabel = styled.label`
	display: inline-flex;
	align-items: center;
	flex-wrap: wrap;
	position: relative;
	user-select: none;
	font-size: ${typography.size.s3}rem;
	font-family: ${typography.type.heading};
`;

const StyledToggleInput = styled.input`
	${visuallyHiddenStyles}

	&:focus + span {
		outline: 2px solid ${colors.emerald75};
		outline-offset: 2px;
	}

	&:checked:not(:disabled) + span {
		background-color: ${colors.emerald};
		border-color: ${colors.emerald};

		&:before {
			transform: translate(1.6rem, -50%);
		}
	}
`;

const StyledToggleSpan = styled.span`
	display: inline-flex;
	align-items: center;
	justify-content: space-around;
	width: 4rem;
	height: 2.4rem;
	border-radius: 1.6rem;
	position: relative;
	background-color: ${colors.grey500};
	border: 1px solid ${colors.grey500};
	cursor: pointer;
	transition: transform 250ms;
	margin-right: 1.7rem;

	@media (prefers-reduced-motion: reduce) {
		transition: none;
	}

	&:before {
		content: '';
		width: 2.2rem;
		height: 2.2rem;
		border-radius: 50%;
		background-color: ${colors.white};
		position: absolute;
		z-index: 2;
		top: 50%;
		left: 0;
		transform: translate(0, -50%);
		transition: inherit;
	}

	&.is-disabled {
		background-color: ${colors.lightGrey};
		border-color: ${colors.lightGrey};
		cursor: not-allowed;

		&:before {
			background-color: ${colors.secondary};
		}
	}
`;

export { StyledToggleLabel, StyledToggleSpan, StyledToggleInput };
