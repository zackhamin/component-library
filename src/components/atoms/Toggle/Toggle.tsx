import cn from 'classnames';
import React from 'react';
import {
	StyledToggleInput,
	StyledToggleLabel,
	StyledToggleSpan,
} from './Toggle.styles';

export interface ToggleProps {
	id: string;
	name?: string;
	children: React.ReactNode;
	isChecked?: boolean;
	isDisabled?: boolean;
	isRequired?: boolean;
	onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

const Toggle = React.forwardRef<HTMLInputElement, ToggleProps>(
	(
		{ id, name, onChange, children, isDisabled, isChecked, isRequired },
		ref
	) => {
		return (
			<StyledToggleLabel htmlFor={id}>
				<StyledToggleInput
					type="checkbox"
					name={name}
					id={id}
					ref={ref}
					disabled={isDisabled}
					onChange={onChange}
					checked={isChecked}
					required={isRequired}
				/>
				<StyledToggleSpan className={cn({ 'is-disabled': isDisabled })} />
				{children} {isRequired && <span aria-label="required">*</span>}
			</StyledToggleLabel>
		);
	}
);

export { Toggle };
