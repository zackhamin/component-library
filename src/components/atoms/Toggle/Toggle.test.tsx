import { Toggle } from '../Toggle';
import userEvent from '@testing-library/user-event';
import { useState } from 'react';
import { render, screen } from '@testing-library/react';

const defaultProps = {
	id: 'toggle',
	name: 'toggle',
};

const ControlledToggle = ({ onChange }: any) => {
	const [checked, setChecked] = useState(false);
	return (
		<Toggle
			{...defaultProps}
			isChecked={checked}
			onChange={(e) => {
				onChange();
				setChecked(e.target.checked);
			}}
		>
			Toggle
		</Toggle>
	);
};

describe('Toggle', () => {
	it('renders correctly', () => {
		const { container } = render(<Toggle {...defaultProps}>Toggle</Toggle>);
		expect(container).toMatchSnapshot();
	});

	it('adds ID & name to element', () => {
		render(<Toggle {...defaultProps}>Toggle</Toggle>);

		const input = screen.getByRole('checkbox');

		expect(input).toHaveAttribute('id', 'toggle');
		expect(input).toHaveAttribute('name', 'toggle');
		expect(input).not.toHaveAttribute('required');
	});
});

describe('Toggle has disabled state', () => {
	it('adds disabled attribute to element', () => {
		render(
			<Toggle {...defaultProps} isDisabled>
				Toggle
			</Toggle>
		);
		expect(screen.getByRole('checkbox')).toHaveAttribute('disabled');
	});

	it('should not check if disabled (uncontrolled)', () => {
		render(
			<Toggle {...defaultProps} isDisabled>
				Toggle
			</Toggle>
		);

		const toggle = screen.getByRole('checkbox');

		expect(toggle).toBeDisabled();
		expect(toggle).not.toBeChecked();

		userEvent.click(toggle);

		expect(toggle).not.toBeChecked();
	});
});

describe('Toggle required state', () => {
	it('should have required attribute', () => {
		render(
			<Toggle {...defaultProps} isRequired>
				Toggle
			</Toggle>
		);

		const toggle = screen.getByRole('checkbox');

		expect(toggle).toHaveAttribute('required');
	});

	it('should have required asterisk', () => {
		render(
			<Toggle {...defaultProps} isRequired>
				Toggle
			</Toggle>
		);

		const toggle = screen.getByLabelText('required');

		expect(toggle).toBeInTheDocument();
	});
});

describe('Toggle change handler', () => {
	it('should check and uncheck (controlled)', () => {
		const onChangeHandler = jest.fn();

		render(<ControlledToggle onChange={onChangeHandler} />);

		const toggle = screen.getByRole('checkbox');

		userEvent.click(toggle);

		expect(onChangeHandler).toHaveBeenCalled();
		expect(toggle).toBeChecked();
	});
});
