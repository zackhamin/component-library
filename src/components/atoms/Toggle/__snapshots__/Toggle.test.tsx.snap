// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle renders correctly 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  font-size: 1.6rem;
  font-family: BrandonGrotesque-Bold,Arial,Helvetica,sans-serif;
}

.c1 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
  white-space: nowrap;
  word-wrap: normal;
}

.c1:focus + span {
  outline: 2px solid #018476;
  outline-offset: 2px;
}

.c1:checked:not(:disabled) + span {
  background-color: #00594C;
  border-color: #00594C;
}

.c1:checked:not(:disabled) + span:before {
  -webkit-transform: translate(1.6rem,-50%);
  -ms-transform: translate(1.6rem,-50%);
  transform: translate(1.6rem,-50%);
}

.c2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  width: 4rem;
  height: 2.4rem;
  border-radius: 1.6rem;
  position: relative;
  background-color: #979797;
  border: 1px solid #979797;
  cursor: pointer;
  -webkit-transition: -webkit-transform 250ms;
  -webkit-transition: transform 250ms;
  transition: transform 250ms;
  margin-right: 1.7rem;
}

.c2:before {
  content: '';
  width: 2.2rem;
  height: 2.2rem;
  border-radius: 50%;
  background-color: #ffffff;
  position: absolute;
  z-index: 2;
  top: 50%;
  left: 0;
  -webkit-transform: translate(0,-50%);
  -ms-transform: translate(0,-50%);
  transform: translate(0,-50%);
  -webkit-transition: inherit;
  transition: inherit;
}

.c2.is-disabled {
  background-color: #EEEEEF;
  border-color: #EEEEEF;
  cursor: not-allowed;
}

.c2.is-disabled:before {
  background-color: #ADB5BD;
}

@media (prefers-reduced-motion:reduce) {
  .c2 {
    -webkit-transition: none;
    transition: none;
  }
}

<div>
  <label
    class="c0"
    for="toggle"
  >
    <input
      class="c1"
      id="toggle"
      name="toggle"
      type="checkbox"
    />
    <span
      class="c2"
    />
    Toggle
     
  </label>
</div>
`;
